lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -4298,7 +4298,11 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     //     print("Caught bar error")
     //   }
     // }
-    if (fromExistential) {
+    //
+    // Note: we relax the restriction if the type we're casting to is a class
+    // because it's possible that we might have a subclass that conforms to
+    // the protocol.
+    if (fromExistential && !toType->hasReferenceSemantics()) {
```

Comment:
There is one simple case we shouldn't relax here: `final class`.

Reply:
Yeah, that makes sense. Let me run the CI to see if anything else is broken and then I'll add another check here to account for a final class.